packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6278,50 +6281,197 @@ abstract class RenderObjectElement extends Element {
+            ),
+            ErrorHint(
+              'This can be resolved by using mixins to separate aspects of a '
+              'ParentData class into separate types, mixing them all in on the '
+              'full ParentData Object, such as KeepAlive does with '
+              'KeepAliveParentDataMixin.'
+            ),
+            ErrorDescription(
+              'The ownership chain for the RenderObject that received the '
+              'parent data was:\n  ${debugGetCreatorChain(10)}'
+            ),
           ]);
-        } on FlutterError catch (e) {
-          _reportException(ErrorSummary('while looking for parent data.'), e, e.stackTrace);
+        } on FlutterError catch (error) {
+          _reportException(
+            ErrorSummary('while looking for parent data.'),
+            error,
+            error.stackTrace,
+          );
         }
       }
       return true;
     }());
+  }
+
+  List<ParentDataElement<ParentData>> _findAncestorParentDataElements() {
+    Element? ancestor = _parent;
+    final List<ParentDataElement<ParentData>> result = <ParentDataElement<ParentData>>[];
+    final Set<Type> debugAncestorTypes = <Type>{};
+    final Set<Type> debugParentDataTypes = <Type>{};
+    final List<Type> debugAncestorCulprits = <Type>[];
+    final List<Type> debugParentDataCulprits = <Type>[];
+
+    // More than one ParentDataWidget can contribute ParentData, but there are
+    // some constraints.
+    // 1. ParentData can only be written by unique ParentDataWidget types.
+    //    For example, two KeepAlive ParentDataWidgets trying to write to the
+    //    same child is not allowed.
+    // 2. Each contributing ParentDataWidget must contribute to a unique
+    //    ParentData type, less ParentData be overwritten.
+    //    For example, there cannot be two ParentDataWidgets that both write
+    //    ParentData of type KeepAliveParentDataMixin, if the first check was
+    //    subverted by a subclassing of the KeepAlive ParentDataWidget.
+    // 3. The ParentData itself must be compatible with all ParentDataWidgets
+    //    writing to it.
+    //    For example, TwoDimensionalViewportParentData uses the
+    //    KeepAliveParentDataMixin, so it could be compatible with both
+    //    KeepAlive, and another ParentDataWidget with ParentData type
+    //    TwoDimensionalViewportParentData or a subclass thereof.
+    // The first and second cases are verified here. The third is verified in
+    // debugIsValidRenderObject.
+
+    while (ancestor != null && ancestor is! RenderObjectElement) {
+      if (ancestor is ParentDataElement<ParentData>) {
+        assert(() {
+          if (!debugAncestorTypes.add(ancestor.runtimeType)) {
+            if (!debugAncestorCulprits.contains(ancestor.runtimeType)) {
+              // Add the first one we had put in the Set of Types.
+              debugAncestorCulprits.add(ancestor.runtimeType);
+            }
+            debugAncestorCulprits.add(ancestor.runtimeType);
```

Comment:
What is this trying to do? It looks like if ancestor.runtimeType is not in the list, it is added twice, if it is in the list it is just added once? Seems odd 😆

Reply:
Yes, tricky huh? 😜
I am adding the first one we did not count as a culprit because we had not seen another one yet. `debugAncestorCulprits` is used to output the list of bad ancestors. After the last review you suggested listing only the culprits instead of all contributing ParentDataWidgets, since more than one is valid now.

Reply:
Oh wait now that I look at it, I reworked this and that logic is no longer necessary. Thanks for catching it!